File: server/auth/currentUser.ts
export class CurrentUser {
  user_id: number
  golden_record_id: string | null
  abilities: string[]
  accountable_institution_id: number
  user_roles_id: number
  tenant_id: string | null

  constructor(data: {
    user_id: number
    golden_record_id?: string | null
    abilities?: string[]
    accountable_institution_id: number
    user_roles_id: number
    tenant_id?: string | null
  }) {
    this.user_id = data.user_id
    this.golden_record_id = data.golden_record_id ?? null
    this.abilities = data.abilities ?? []
    this.accountable_institution_id = data.accountable_institution_id
    this.user_roles_id = data.user_roles_id
    this.tenant_id = data.tenant_id ?? null
  }

  hasAbility(ability: string): boolean {
    return this.abilities.includes(ability)
  }

  isRole(...roles: number[]): boolean {
    return roles.includes(this.user_roles_id)
  }

  /**
   * user_roles_id === 1 (Super Admin) — documented in Transfers handover §4.3.
   */
  get isSuperAdmin(): boolean {
    return this.user_roles_id === 1
  }

  /**
   * user_roles_id === 4 (Client) — documented in Transfers handover §4.3.
   */
  get isClient(): boolean {
    return this.user_roles_id === 4
  }

  toJSON() {
    return {
      user_id: this.user_id,
      golden_record_id: this.golden_record_id,
      abilities: this.abilities,
      accountable_institution_id: this.accountable_institution_id,
      user_roles_id: this.user_roles_id,
      tenant_id: this.tenant_id,
    }
  }
}

File: python_server/auth/current_user.py
import uuid
from dataclasses import dataclass
from typing import List, Optional


@dataclass(frozen=True)
class CurrentUser:
    user_id: int
    golden_record_id: Optional[uuid.UUID]
    abilities: List[str]
    accountable_institution_id: int
    user_roles_id: int
    tenant_id: Optional[uuid.UUID]

    def has_ability(self, ability: str) -> bool:
        return ability in self.abilities

    def is_role(self, *roles: int) -> bool:
        return self.user_roles_id in roles

    @property
    def is_super_admin(self) -> bool:
        return self.user_roles_id == 1

    @property
    def is_client(self) -> bool:
        return self.user_roles_id == 4

    @property
    def is_agent_or_above(self) -> bool:
        return self.user_roles_id in (1, 2, 3, 5, 6)

    @property
    def can_approve_high_risk(self) -> bool:
        return self.user_roles_id in (1, 2, 5)

    def __repr__(self) -> str:
        return (
            f"CurrentUser(user_id={self.user_id}, "
            f"accountable_institution_id={self.accountable_institution_id}, "
            f"user_roles_id={self.user_roles_id}, "
            f"abilities=[...])"
        )
